### PR TITLE
e2e: add NodeLogQuery tests for Windows nodes

### PR DESCRIPTION
**Pull request by aravindhp**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Add e2e tests for the NodeLogQuery feature against Windows nodes

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

File: test/e2e/node/kubelet.go [hunk 1]
```diff
@@ -611,6 +611,31 @@ var _ = SIGDescribe("kubelet", func() {
...
+			e2eskipper.SkipUnlessProviderIs(framework.ProvidersWithSSH...)
+			if len(windowsNodeName) == 0 {
+				ginkgo.Skip("No Windows node found")
+			}
+			ginkgo.By("Starting the command")
+			tk := e2ekubectl.NewTestKubeconfig(framework.TestContext.CertDir, framework.TestContext.Host, framework.TestContext.KubeConfig, framework.TestContext.KubeContext, framework.TestContext.KubectlPath, ns)
+
+			queryCommand := fmt.Sprintf("/api/v1/nodes/%s/proxy/logs/?query=Microsoft-Windows-Security-SPP&tailLines=3", windowsNodeName)
+			cmd := tk.KubectlCmd("get", "--raw", queryCommand)
+			result := runKubectlCommand(cmd)
```
Comment (aravindhp) [inline test/e2e/node/kubelet.go:629]: This shows up as:
```
  Jan  5 16:15:49.155: INFO: Kubectl output: 
  
     ProviderName: Microsoft-Windows-Security-SPP
  
  TimeCreated             Id LevelDisplayName Message                                                                    
  -----------             -- ---------------- -------                                                                    
  1/5/2024 11:30:44 PM 16384 Information      Successfully scheduled Software Protection service for re-start at         
                                              2024-01-12T22:32:44Z. Reason: RulesEngine.                                 
  1/5/2024 11:36:15 PM 16394 Information      Offline downlevel migration succeeded.                                     
  1/5/2024 11:36:45 PM 16384 Information      Successfully scheduled Software Protection service for re-start at         
                                              2024-01-12T22:32:45Z. Reason: RulesEngine.                                 
  
  

```

File: test/e2e/node/kubelet.go [hunk 2]
```diff
@@ -670,10 +695,20 @@ func assertContains(expectedString string, result string) {
...
 	e2essh.LogResult(result)
 	return result.Stdout
 }
+
+func journalctlCommandOnNode(nodeName string, args string) string {
+	return commandOnNode(nodeName, "journalctl --utc --no-pager --output=short-precise "+args)
+}
+
+func getWinEventCommandOnNode(nodeName string, providerName, args string) string {
+	output := commandOnNode(nodeName, "Get-WinEvent -FilterHashtable @{LogName='Application'; ProviderName='"+providerName+"'}"+args+" | Format-Table -AutoSize -Wrap")
```
Comment (aravindhp) [inline test/e2e/node/kubelet.go:710]: This shows up as:
```
  Jan  5 16:15:51.164: INFO: ssh capi@capz-conf-2c5l6:22: stdout:    "\r\n\r\n   ProviderName: Microsoft-Windows-Security-SPP\r\n\r\nTimeCreated             Id LevelDisplayName Message                                                \r\n-----------             -- ---------------- -------                                                \r\n1/5/2024 11:36:45 PM 16384 Information      Successfully scheduled Software Protection service for \r\n                                            re-start at 2024-01-12T22:32:45Z. Reason: RulesEngine. \r\n1/5/2024 11:36:15 PM 16394 Information      Offline downlevel migration succeeded.                 \r\n1/5/2024 11:30:44 PM 16384 Information      Successfully scheduled Software Protection service for \r\n                                            re-start at 2024-01-12T22:32:44Z. Reason: RulesEngine. \r\n\r\n\r\n"
```

File: test/e2e/node/kubelet.go [hunk 3]
```diff
@@ -670,10 +695,20 @@ func assertContains(expectedString string, result string) {
...
+	// Attempt to normalize the output of Get-WinEvent
+	output = strings.Replace(output, "\r\n", "\n", -1)
```
Comment (aravindhp) [inline test/e2e/node/kubelet.go:712]: After the attempt to replace the Windows newline with Linux newlines, this shows up as:

```
  Jan  5 16:15:51.165: INFO: Get-WinEvent output: 

     ProviderName: Microsoft-Windows-Security-SPP

  TimeCreated             Id LevelDisplayName Message                                                
  -----------             -- ---------------- -------                                                
  1/5/2024 11:36:45 PM 16384 Information      Successfully scheduled Software Protection service for 
                                              re-start at 2024-01-12T22:32:45Z. Reason: RulesEngine. 
  1/5/2024 11:36:15 PM 16394 Information      Offline downlevel migration succeeded.                 
  1/5/2024 11:30:44 PM 16384 Information      Successfully scheduled Software Protection service for 
                                              re-start at 2024-01-12T22:32:44Z. Reason: RulesEngine. 


```

But if you compare with the kubelet output, you can see the newlines are slightly out of whack. Not sure how to normalize this...

**Review by aravindhp — COMMENTED** [review]
@jsturtevant @marosset any ideas on how to normalize the newlines across Windows and Linux for comparison purposes?

File: test/e2e/node/kubelet.go (hunk 3 above)
Comment (jsturtevant) [reply test/e2e/node/kubelet.go:712]: I am not sure I understand what I am comparing this too and what is "out of whack".  What are we trying to do?

Comment (aravindhp) [reply test/e2e/node/kubelet.go:712]: Apologies for not being clear. I am comparing the above output with the output in [this comment](https://github.com/kubernetes/kubernetes/pull/122594#discussion_r1443536128). I am trying to confirm that the output returned by kubectl is the same as the output returned by Get-WinEvent executed on the node.

Comment (jsturtevant) [reply test/e2e/node/kubelet.go:712]: it looks like the powershell commands are not in the same order (one is descending the other is ascending) and the line breaks are different places. What is the command formatting being run on kubelet?  

For ordering, we might need to ensure we are  ordering them properly (something like `Sort-Object -Property TimeCreated -Descending`)

For the line breaks we probably cannot use `Format-Table -AutoSize`.  It seems like you are getting different results depending on if run from ssh or directly by kubelet.

Comment (jsturtevant) [reply test/e2e/node/kubelet.go:712]: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.diagnostics/get-winevent?view=powershell-7.4#description `By default, Get-WinEvent returns event information in the order of newest to oldest.`  so maybe default is being changed somewhere?

**Comment by kannon92** [comment]
PR looks to be related to https://github.com/kubernetes/enhancements/issues/2258.

Its unclear to me at what stage this feature is by the issue.

/triage accepted
/priority important-soon

We should get these e2e tests in this release as this feature is currently alpha (I think!).

**Comment by aravindhp** [comment]
@kannon92 
> PR looks to be related to [kubernetes/enhancements#2258](https://github.com/kubernetes/enhancements/issues/2258).

That is correct

> Its unclear to me at what stage this feature is by the issue.

Feature is in alpha. Planning to graduate to beta in 1.30.

> We should get these e2e tests in this release as this feature is currently alpha (I think!).

That is what I am working on. We already have periodic e2e tests running for Linux.

File: test/e2e/node/kubelet.go (hunk 3 above)
Comment (aravindhp) [reply test/e2e/node/kubelet.go]: This issue was resolved by normalizing both sides of the output by removing the newline characters.

**Comment by kannon92** [comment]
Thank you @aravindhp. I was commented on it as I was traging the feature and decided the priority.

**Comment by aravindhp** [comment]
/retitle  e2e: add NodeLogQuery tests for Windows nodes

**Comment by aravindhp** [comment]
@marosset @jsturtevant please review

**Review by mrunalp — APPROVED** [review]

**Pull request merged**